### ISSUE TITLE
Release the demonstration/grasp collection code

### ISSUE THREAD

**Issue by williamshen-nz** [f3rm/f3rm#6]
See: https://github.com/f3rm/f3rm/issues/4#issuecomment-1744692600

Need to test out VR support @geyang

**Comment by Gloryseven** [comment]
how can I make my owe datesets? The robotic manipulation needs a  Franka Panda gripper ,but I do not have this machine. Does It mean that I can not test the robotic manipulation by myself?

**Comment by williamshen-nz** [comment]
If you already have a robot with a calibrated camera, you can use forward kinematics to compute the camera pose and use that to create your own `transforms.json` file of images captured at different viewpoints.

If you just want to try out the feature fields side (no robot optimization), you can create your own datasets using the instructions from Nerfstudio: https://docs.nerf.studio/quickstart/custom_dataset.html. Training a feature field and interacting with it through the visualizer does not require a robot.

The robot optimization code assumes:
1. You know the transformation to get from the NeRF coordinate frame to the world coordinate frame (i.e., the robot's coordinate system) so we can limit the workspace of the optimization and ensure correct scale. See https://github.com/f3rm/f3rm/issues/7
2. You are using a Franka Panda gripper collision checking purposes. You can check the instructions here for how to modify it to suit your robot platform: https://github.com/f3rm/f3rm/tree/main/f3rm_robot#what-if-im-using-a-different-gripper-or-robot

**Comment by Gloryseven** [comment]
Hello! I would like to ask where the code of learning and testing for ‘Few-Shot Manipulation’ section in the paper is located.

**Comment by williamshen-nz** [comment]
Hi, you should be able to easily implement the few-shot manipulation method by modifying the existing [optimize.py](https://github.com/f3rm/f3rm/blob/main/f3rm_robot/optimize.py) script. The main changes you will need to make are removing the language guidance in the loss function, and using different masking to generate initial proposals now that you do not have language. The details of how to do so and the hyperparameters we use can be found in the paper.

**Comment by williamshen-nz** [comment]
PR in progress here: https://github.com/f3rm/f3rm/pull/13

You can `pip install vuer` to start playing around with it

**Comment by Gloryseven** [comment]
hello! how could I create the 'scene_demo*.json' file on my own demo datasets?

**Comment by williamshen-nz** [comment]
Hi, you can teleop the robot in the real-world to the pose you want to collect the demonstration at. Note that the demonstrations are provided in the world coordinate frame.

You just need to follow the schema. The example files should tell you what you need. Our labeling system was working in the browser but not VR last I tried, so you could try pull the branch in the PR and using [this script](https://github.com/f3rm/f3rm/blob/c88a6be80e7f9ceb7276b10ff5fe52b659ef0510/f3rm_robot/examples/label_demos.py).

**Comment by Gloryseven** [comment]
Have you conducted experiments in a simulation environment? (All of your datasets are in real scenarios.)

**Comment by williamshen-nz** [comment]
No we didn't, all the experiments were conducted in the real world. You might find it difficult to get a good quality NeRF unless you use a simulator with high quality rendering and assets with good texture, etc.

**Comment by KamiPan0601** [comment]
Hi, how can I use the `dataparser_transforms.json` to compute the inverse transformation (Is there a formula) ？ And how can I get `nerf_to_world.json` with the existing `transforms.json`？I come from #7 and still have this problem. Thank you :)

> If you already have a robot with a calibrated camera, you can use forward kinematics to compute the camera pose and use that to create your own `transforms.json` file of images captured at different viewpoints.
> 
> If you just want to try out the feature fields side (no robot optimization), you can create your own datasets using the instructions from Nerfstudio: https://docs.nerf.studio/quickstart/custom_dataset.html. Training a feature field and interacting with it through the visualizer does not require a robot.
> 
> The robot optimization code assumes:
> 
> 1. You know the transformation to get from the NeRF coordinate frame to the world coordinate frame (i.e., the robot's coordinate system) so we can limit the workspace of the optimization and ensure correct scale. See [Nerf_to_world.json #7](https://github.com/f3rm/f3rm/issues/7)

**Comment by williamshen-nz** [comment]
@KamiPan0601 I have added some more details here which may help: #15 

For computing the inverse transformation, it would be the opposite of what nerfstudio is doing to compute the forward transformation. So you will need to apply the inverse scale, rotation and translation. You can check how they do it here: https://github.com/nerfstudio-project/nerfstudio/blob/v0.3.4/nerfstudio/data/dataparsers/nerfstudio_dataparser.py#L233-L244

**Comment by dajiebi** [comment]
@KamiPan0601 Has the problem been solved? 
I had the same problem as you. Now that I have transform and transform_gt and have computed colmap_to_world,
 however, how do I compute nerf_to_colmap from dataparser_transforms.json?
 How do I combine colmap_to_world (T,scale) and nerf_to_colmap (which seems to be the inverse of dataparser_transforms.json, dataparser_transforms.json only has a 3*4 matrix and a scale. How do I construct nerf_to_colmap)?

**Comment by KamiPan0601** [comment]
@dajiebi Yes, it's been solved. Also suggest you read the source code above. 
First apply the inverse scale to the translation part of the 3\*4 matrix (last column), by doing the opposite of what the following code does. (I use `ns-train f3rm --data 'data_path' nerfstudio-data --auto-scale-poses False` so my scale_factor = 1.0)
[https://github.com/nerfstudio-project/nerfstudio/blob/v0.3.4/nerfstudio/data/dataparsers/nerfstudio_dataparser.py#L246](url)
Then turn the 3\*4 matrix into a 4\*4 homogeneous transformation matrix by adding a row [0 0 0 1] at the end. Finally, Compute its inverse matrix.
If you use this nerf_to_colmap matrix as the `nerf_to_world.json` to optimize, the grasp pose you get is under the colmap coordinate system. Then you need to transform the pose back to the world coordinate with colmap_to_world.

**Comment by dajiebi** [comment]
Does "transform the pose back to the world coordinate with colmap_to_world" mean `colmap_to_world @ grasp`?

**Comment by dajiebi** [comment]
When I use `ns-train f3rm --data 'data_path' nerfstudio-data --auto-scale-poses False` ,nerfstudio can not output a good reconstruction result,How did you solve this problem?

**Comment by williamshen-nz** [comment]
This means that your unscaled poses are not amenable for training a NeRF. This could be because your poses are all outside the unit cube or they are very constrained within it.

You should check the poses in the nerfstudio viewer. To solve the problem you need to rescale the poses, which you can do using `--auto-scale-poses True`. When computing your NeRF to COLMAP matrix, you will need to use the computed scale.

**Comment by dajiebi** [comment]
Thank you for your reply! Is `nerf_to_world` just `colmap_to_world`@ np.linalg.inv (`scale_matrix` @ `dataparse_transform`)?

**Comment by williamshen-nz** [comment]
Please use Vuer to collect demonstrations in VR: https://github.com/vuer-ai/vuer

Example of teleopping humanoids here: https://github.com/OpenTeleVision/TeleVision
